File: samples/cdk/typescript/quickstartDiscovery.ts
/**
 * Quickstart: Implement a simple health discovery connector.
 *
 * Demonstrates setting up HTTP health probing for a provider and
 * retrieving health reports. This quickstart shows a minimal inline
 * implementation of the DiscoveryConnector interface.
 *
 * For a full YAML-registry-based implementation with rolling availability
 * scores, see samples/connectors/typescript/customDiscovery.ts.
 */

import {
    ProbeResult,
    HealthReport,
    DiscoveryConnector,
    SyncResult,
    SyncStatus,
} from "@modelmesh/core";

/**
 * Simple health probe discovery connector.
 *
 * Registers provider endpoints and simulates health probes.
 * A real implementation would make HTTP requests to the provider's
 * health endpoint (e.g., /v1/models, /health, /readiness).
 */
class SimpleHealthDiscovery implements DiscoveryConnector {
    private endpoints = new Map<string, string>();

    registerProviderUrl(providerId: string, baseUrl: string): void {
        this.endpoints.set(providerId, baseUrl);
    }

    async sync(_providers?: string[]): Promise<SyncResult> {
        return { newModels: [], deprecatedModels: [], updatedModels: [], errors: [] };
    }

    async getSyncStatus(): Promise<SyncStatus> {
        return { modelsSynced: 0, status: "idle" };
    }

    async probe(providerId: string): Promise<ProbeResult> {
        const baseUrl = this.endpoints.get(providerId);
        if (!baseUrl) {
            return { providerId, success: false, error: `Unknown provider: ${providerId}` };
        }
        // In a real implementation, make an HTTP GET to the health endpoint
        // and measure latency. For this quickstart, we return a simulated result.
        return {
            providerId,
            success: true,
            latencyMs: 42.5,
            statusCode: 200,
        };
    }

    async getHealthReport(providerId?: string): Promise<HealthReport[]> {
        const ids = providerId ? [providerId] : [...this.endpoints.keys()];
        return ids.map(id => ({
            providerId: id,
            available: true,
            latencyMs: 42.5,
            statusCode: 200,
            availabilityScore: 1.0,
            timestamp: new Date(),
        }));
    }
}

async function main(): Promise<void> {
    const discovery = new SimpleHealthDiscovery();

    // Register the provider URL for probing
    discovery.registerProviderUrl("openai", "https://api.openai.com");

    // Probe the provider
    const result: ProbeResult = await discovery.probe("openai");
    console.log(`Probe success: ${result.success}`);
    console.log(`Latency: ${result.latencyMs?.toFixed(1)}ms`);

    // Get a full health report
    const reports: HealthReport[] = await discovery.getHealthReport("openai");
    for (const report of reports) {
        console.log(`Provider: ${report.providerId}, Available: ${report.available}`);
        console.log(`Availability score: ${report.availabilityScore.toFixed(2)}`);
    }
}

main().catch(console.error);

File: samples/cdk/python/quickstart_discovery.py
"""Quickstart: Create an HttpHealthDiscovery, probe a provider, get health report.

Demonstrates setting up HTTP health probing for a provider and retrieving
health reports using the HttpHealthDiscovery specialized class.
"""

import asyncio

from modelmesh.cdk import HttpHealthDiscovery, HttpHealthDiscoveryConfig


async def main() -> None:
    discovery = HttpHealthDiscovery(HttpHealthDiscoveryConfig(
        providers=["openai"],
        health_path="/v1/models",
        expected_status=200,
        health_interval_seconds=30,
        failure_threshold=3,
    ))

    # Register the provider URL for probing
    discovery.register_provider_url("openai", "https://api.openai.com")

    # Probe the provider
    result = await discovery.probe("openai")
    print(f"Probe success: {result.success}")
    print(f"Latency: {result.latency_ms:.1f}ms")

    # Get a full health report
    reports = await discovery.get_health_report("openai")
    for report in reports:
        print(f"Provider: {report.provider_id}, Available: {report.available}")
        print(f"Availability score: {report.availability_score:.2f}")

    # No close() needed -- HttpHealthDiscovery has no persistent connections


if __name__ == "__main__":
    asyncio.run(main())
